docs/implementation.html

```diff
@@ -102,6 +102,56 @@ <h5><a id="recordheader" href="#recordheader">5.4.2.1 Record Header</a></h5>
     is also encoded as a varint.</p>
 
+    <h4><a id="messageset" href="#messageset">5.3.3 Old Message Format</a></h4>
+    <p>
+        Prior to Kafka 0.11 messages were transferred and stored in a format called MessageSets. In MessageSets each message has its own metadata. Note that although MessageSets are formally arrays,
+        they are not preceded by an int32 array size like other array elements in the protocol.
+    </p>
+
+    <b>Message Set:</b><br>
+    <p><pre class="brush: java;">
+    MessageSet (Version: 0) => [offset message_size message]
+        offset => INT64
+        message_size => INT32
+        message => crc magic_byte attributes key value
+            crc => INT32
+            magic_byte => INT8
+            attributes => INT8
+            key => BYTES
+            value => BYTES
+    </pre></p>
+    <p><pre class="brush: java;">
+    MessageSet (Version: 1) => [offset message_size message]
+        offset => INT64
+        message_size => INT32
+        message => crc magic_byte attributes key value
+            crc => INT32
+            magic_byte => INT8
+            attributes => INT8
+            timestamp =>INT64
+            key => BYTES
+            value => BYTES
+    </pre></p>
+    <p>
+        In versions prior to Kafka 0.10 the magic byte value was 0. Message format version 1 was introduced with timestamp support in version 0.10 .
+    <ul>
+        <li>Similarly to version 2 above, the lowest bits of attributes mean compression type.</li>
+        <li>In version 1 the fourth lowest bit of attributes byte represents the timestamp type.<br>
+            0 stands for CreateTime and 1 stands for LogAppendTime. The producer should always set this bit to 0 and if the topic is configured to contain log append times
+            (through either broker level config log.message.timestamp.type = LogAppendTime or topic level config message.timestamp.type = LogAppendTime), the broker will overwrite the timestamp type and the timestamp in the message set.</li>
+
+        <li>The highest bits of attributes must be set to 0.</li>
+    </ul>
+    </p>
+    <p>In message format versions 0 and 1 Kafka supports recursive messages for efficiency. In this case the message's attributes show that it is compressed and its value field contains a message set compressed with the appropriate compression type.</p>
+    <p>
+        When receiving recursive version 0 messages, the broker decompresses them and each inner message is assigned an offset.
+        In version 1, to avoid server side re-compression, each compressed message must have offset starting from 0 and increasing by one for each inner message in the compressed message.
+        When the producer is sending non compressed messages, it can set the offsets to anything.
+    </p>
+
+    <p>The crc field contains the CRC32 (and not CRC-32C) of the message bytes.</p>
```

Comment:
Maybe worthwhile emphasizing that the CRC covers all fields from the magic to the message value?